patches/tock/10-connect-vendor-hid-usb-interface.patch

```diff
@@ -0,0 +1,349 @@
++    // 64-byte buffers for the endpoints.
++    // Endpoints are numbered from 1, but indexed here from 0.
++    in_buffer: [Buffer64; NUM_ENDPOINTS],
++    out_buffer: [Buffer64; NUM_ENDPOINTS],
+ 
+     // Interaction with the client
+     client: OptionalCell<&'b dyn CtapUsbClient>,
+     tx_packet: OptionalCell<[u8; 64]>,
+     pending_in: Cell<bool>,
+     pending_out: Cell<bool>,
+-    delayed_out: Cell<bool>,
++    // The endpoint of the delayed packet.
++    delayed_out: OptionalCell<usize>,
+ }
+ 
+ impl<'a, 'b, C: hil::usb::UsbController<'a>> ClientCtapHID<'a, 'b, C> {
+@@ -167,7 +192,7 @@ impl<'a, 'b, C: hil::usb::UsbController<'a>> ClientCtapHID<'a, 'b, C> {
+             &[
+                 EndpointDescriptor {
+                     endpoint_address: EndpointAddress::new_const(
+-                        ENDPOINT_NUM + 1,
++                        VENDOR_ENDPOINT_NUM,
+                         TransferDirection::HostToDevice,
+                     ),
+                     transfer_type: TransferType::Interrupt,
+@@ -176,7 +201,7 @@ impl<'a, 'b, C: hil::usb::UsbController<'a>> ClientCtapHID<'a, 'b, C> {
+                 },
+                 EndpointDescriptor {
+                     endpoint_address: EndpointAddress::new_const(
+-                        ENDPOINT_NUM + 1,
++                        VENDOR_ENDPOINT_NUM,
+                         TransferDirection::DeviceToHost,
+                     ),
+                     transfer_type: TransferType::Interrupt,
+@@ -228,13 +253,22 @@ impl<'a, 'b, C: hil::usb::UsbController<'a>> ClientCtapHID<'a, 'b, C> {
+                 LANGUAGES,
+                 strings,
+             ),
+-            in_buffer: Buffer64::default(),
+-            out_buffer: Buffer64::default(),
++            in_buffer: [
++                Buffer64::default(),
++                #[cfg(feature = "vendor_hid")]
++                Buffer64::default(),
++            ],
++            out_buffer: [
++                Buffer64::default(),
++                #[cfg(feature = "vendor_hid")]
++                Buffer64::default(),
++            ],
+             client: OptionalCell::empty(),
+             tx_packet: OptionalCell::empty(),
+             pending_in: Cell::new(false),
+             pending_out: Cell::new(false),
+-            delayed_out: Cell::new(false),
++            delayed_out: OptionalCell::empty(),
++
+         }
+     }
+ 
+@@ -242,22 +276,26 @@ impl<'a, 'b, C: hil::usb::UsbController<'a>> ClientCtapHID<'a, 'b, C> {
+         self.client.set(client);
+     }
+ 
+-    pub fn transmit_packet(&'a self, packet: &[u8]) -> bool {
++    pub fn transmit_packet(&'a self, packet: &[u8], endpoint: usize) -> bool {
+         if self.pending_in.get() {
+             // The previous packet has not yet been transmitted, reject the new one.
+             false
+         } else {
++            if endpoint_to_buffer(endpoint).is_none() {
++                // Unsupported endpoint
++                return false;
```

Comment:
This is not the best error. The user will get `EBUSY` instead of `EINVAL`. I would have just checked `endpoint` directly in the command (which means exposing a function to know the number of endpoints). But if that's too complicated, I guess this version is acceptable.

Reply:
Changed the return code to be the status code we want, rather than a bool.